app/services/editor-commands/editor-commands.ts

```diff
@@ -0,0 +1,273 @@
+import { mutation, StatefulService } from 'services/core/stateful-service';
+import { Command } from './commands/command';
+import * as commands from './commands';
+import { CombinableCommand } from './commands/combinable-command';
+import { shortcut } from 'services/shortcuts';
+import { SelectionService } from 'services/selection';
+import { Inject } from 'services/core/injector';
+import { ENudgeDirection } from './commands/nudge-items';
+import { SceneCollectionsService } from 'services/scene-collections';
+import electron from 'electron';
+
+const COMMANDS = { ...commands };
+
+const COMBINE_TIMEOUT = 500;
+
+interface ICommandMetadata {
+  description: string;
+}
+
+interface IEditorCommandsServiceState {
+  undoMetadata: ICommandMetadata[];
+  redoMetadata: ICommandMetadata[];
+  operationInProgress: boolean;
+}
+
+export class EditorCommandsService extends StatefulService<IEditorCommandsServiceState> {
+  @Inject() private selectionService: SelectionService;
+  @Inject() private sceneCollectionsService: SceneCollectionsService;
+
+  static initialState: IEditorCommandsServiceState = {
+    undoMetadata: [],
+    redoMetadata: [],
+    operationInProgress: false,
+  };
+
+  // IMPORTANT: Must be kept in sync with metadata stored in
+  // this service's vuex state.
+  undoHistory: Command[] = [];
+  redoHistory: Command[] = [];
```

Comment:
not sure why this isnt kept in state in the first place

Reply:
They are complex classes that can't be serialized.